### PR TITLE
Implemented cache decorator!

### PR DESCRIPTION
**Pull request by raychw**
Implemented cache decorator!

### CODE REVIEW

File: app/main.py [hunk 1]
```diff
@@ -1,3 +1,19 @@
-def cache(func):
-    # Write your code here
-    pass
+from typing import Callable
+
+
+def cache(func: Callable) -> Callable:
+    cached_results = {}
+
+    def wrapper(*args) -> Callable:
```
Comment (Polyakiv-Andrey) [inline app/main.py:7]: The video tutorials talked about a tool that allows you not to lose the name of the decorated function and its docstring. This tool is in the functools library.

**Review by Polyakiv-Andrey — CHANGES_REQUESTED** [review]

File: app/main.py [hunk 2]
```diff
@@ -1,3 +1,21 @@
...
+from typing import Callable
+from functools import wraps
+
+
+def cache(func: Callable) -> Callable:
+    cached_results = {}
+
+    @wraps(func)
+    def wrapper(*args) -> Callable:
+        key = args
```
Comment (Dimosphen1) [inline app/main.py:10]: Do not assign one variable to another, you can operate with `args` directly.

File: app/main.py [hunk 3]
```diff
@@ -1,3 +1,21 @@
...
+
+        if key in cached_results:
+            print("Getting from cache")
+            return cached_results[key]
+        else:
```
Comment (Dimosphen1) [inline app/main.py:15]: There is no need for the `else` statement, code will not execute after `return`.

File: app/main.py [hunk 4]
```diff
@@ -1,3 +1,21 @@
...
+            print("Calculating new result")
+            result = func(*args)
+            cached_results[key] = result
```
Comment (Dimosphen1) [inline app/main.py:18]: There is no need for a `result` variable, write function output to the cache directly.

File: app/main.py [hunk 5]
```diff
@@ -1,3 +1,21 @@
-def cache(func):
-    # Write your code here
-    pass
+from typing import Callable
+from functools import wraps
```
Comment (Dimosphen1) [inline app/main.py:2]: You should perform imports in alphabetical order here.

File: app/main.py [hunk 6]
```diff
@@ -1,3 +1,21 @@
...
+
+
+def cache(func: Callable) -> Callable:
+    cached_results = {}
+
+    @wraps(func)
+    def wrapper(*args) -> Callable:
```
Comment (Dimosphen1) [inline app/main.py:9]: The return type of `wrapper` can be any, because you return the result of calculations, not the function itself.

**Review by Dimosphen1 — CHANGES_REQUESTED** [review]
Several changes were requested.

**Review by Polyakiv-Andrey — APPROVED** [review]
Good job!

**Pull request closed without merging**